File: tests/oracle/seqnum_identity_1204.cc
#include "bfd_arch.hh"
#include "graph.hh"
#include "libdecomp.hh"
#include "unionresolve.hh"
#include "variable.hh"

#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::string;
using std::vector;

string opTimes(Funcdata *fd)
{
  string result;
  for(PcodeOpTree::const_iterator iter=fd->beginOpAll();iter!=fd->endOpAll();++iter) {
    if (!result.empty()) result += ',';
    result += std::to_string((*iter).second->getTime());
  }
  return result;
}

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary,"default",&std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found");

    AddrSpace *codeSpace = architecture.getDefaultCodeSpace();
    AddrSpace *registerSpace = architecture.getSpaceByName("register");
    BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
    BlockBasic *block = graph.newBlockBasic(fd);

    PcodeOp *first = fd->newOp(0,Address(codeSpace,0x1000));
    PcodeOp *middle = fd->newOp(0,Address(codeSpace,0x1000));
    PcodeOp *last = fd->newOp(0,Address(codeSpace,0x1000));
    fd->opSetOpcode(first,CPUI_COPY);
    fd->opSetOpcode(middle,CPUI_COPY);
    fd->opSetOpcode(last,CPUI_COPY);
    Varnode *firstOut = fd->newVarnodeOut(4,Address(registerSpace,0x20),first);
    Varnode *middleOut = fd->newVarnodeOut(4,Address(registerSpace,0x20),middle);

    const uintm firstTime = first->getTime();
    const uintm middleTime = middle->getTime();
    const SeqNum firstIdentity = first->getSeqNum();
    const SeqNum middleIdentity = middle->getSeqNum();
    const SeqNum sameTimeOtherAddress(Address(codeSpace,0x2000),firstTime);
    fd->opInsertEnd(first,block);
    fd->opInsertEnd(middle,block);
    fd->opInsertEnd(last,block);

    // Repeated insertions into the same shrinking gap force BlockBasic::setOrder
    // more than once. SeqNum time/identity and both bank keys must remain stable.
    for(int4 i=0;i<70;++i) {
      PcodeOp *padding = fd->newOp(0,Address(codeSpace,0x1000));
      fd->opSetOpcode(padding,CPUI_COPY);
      fd->opInsertBefore(padding,middle);
    }
    fd->opUninsert(first);
    fd->opInsertEnd(first,block); // time(first)<time(middle), order(first)>order(middle)

    PcodeOp *firstLookup = fd->findOp(firstIdentity);
    PcodeOp *middleLookup = fd->findOp(middleIdentity);
    VarnodeLocSet::const_iterator firstVn = fd->beginLoc(4,firstOut->getAddr(),first->getAddr(),firstTime);
    VarnodeLocSet::const_iterator middleVn = fd->beginLoc(4,middleOut->getAddr(),middle->getAddr(),middleTime);

    std::cout << "time_order=" << first->getTime() << ':' << first->getSeqNum().getOrder()
              << ',' << middle->getTime() << ':' << middle->getSeqNum().getOrder() << '\n';
    std::cout << "identity=" << (firstIdentity == first->getSeqNum())
              << ',' << (middleIdentity == middle->getSeqNum()) << '\n';
    std::cout << "cross_address=" << (firstIdentity == sameTimeOtherAddress)
              << ',' << (firstIdentity < sameTimeOtherAddress) << '\n';
    std::cout << "lookup=" << (firstLookup == first) << ',' << (middleLookup == middle) << '\n';
    std::cout << "varnode_lookup=" << (firstVn != fd->endLoc() && *firstVn == firstOut)
              << ',' << (middleVn != fd->endLoc() && *middleVn == middleOut) << '\n';
    std::cout << "optree_prefix=" << opTimes(fd).substr(0,5) << '\n';
    std::cout << "order_relation="
              << (first->getSeqNum().getOrder() < middle->getSeqNum().getOrder()) << '\n';

    std::ostringstream graphDump;
    dump_dataflow_graph(*fd,graphDump);
    const string graphText = graphDump.str();
    const string opSection = graphText.substr(
      graphText.find("//START:opnodes"),
      graphText.find("*END_COLUMNS",graphText.find("//START:opnodes"))
        - graphText.find("//START:opnodes"));
    std::cout << "graph_time="
              << (opSection.find("\no" + std::to_string(firstTime) + " ") != string::npos)
              << ',' << (graphText.find("\no" + std::to_string(firstTime) + " v") != string::npos)
              << '\n';

    Datatype *parentType = architecture.types->getBase(4,TYPE_UNKNOWN);
    ResolveEdge firstSlotOne(parentType,first,1);
    ResolveEdge middleSlotZero(parentType,middle,0);
    ResolveEdge firstSlotZero(parentType,first,0);
    std::cout << "resolve_order=" << (firstSlotOne < middleSlotZero)
              << ',' << (firstSlotZero < middleSlotZero) << '\n';
    std::cout << "compare_name=" << HighVariable::compareName(firstOut,middleOut)
              << ',' << HighVariable::compareName(middleOut,firstOut) << '\n';
    fd->opDestroy(first);
    VarnodeLocSet::const_iterator deleted = fd->beginLoc(4,Address(registerSpace,0x20),
                                                         Address(codeSpace,0x1000),firstTime);
    VarnodeLocSet::const_iterator retained = fd->beginLoc(4,Address(registerSpace,0x20),
                                                          Address(codeSpace,0x1000),middleTime);
    std::cout << "destroy_identity="
              << (deleted == fd->endLoc() || *deleted != firstOut)
              << ',' << (retained != fd->endLoc() && *retained == middleOut) << '\n';
  }
  shutdownDecompilerLibrary();
}

} // namespace

int main(int argc,char **argv)
{
  if (argc != 3) return 2;
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const std::exception &err) {
    std::cerr << err.what() << '\n';
    return 1;
  }
}

File: tests/oracle/seqnum_identity_1204.rs
use rugra::address::{Address, SeqNum};
use rugra::block::{BlockBasic, FlowBlock};
use rugra::funcdata::Funcdata;
use rugra::graph::dump_dataflow_graph_string;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::{Datatype, TypeBase, TypeMetatype};
use rugra::unionresolve::ResolveEdge;
use rugra::variable::HighVariable;
use std::sync::{Arc, RwLock};

fn main() {
    let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 74);
    let block: Arc<RwLock<dyn FlowBlock + Send + Sync>> =
        Arc::new(RwLock::new(BlockBasic::new(0, Address::new(0x1000))));
    fd.bblocks.add_block(block.clone());

    let first = fd.new_op(0, Address::new(0x1000));
    let middle = fd.new_op(0, Address::new(0x1000));
    let last = fd.new_op(0, Address::new(0x1000));
    fd.op_set_opcode(&first, OpCode::CPUI_COPY);
    fd.op_set_opcode(&middle, OpCode::CPUI_COPY);
    fd.op_set_opcode(&last, OpCode::CPUI_COPY);
    let first_out = fd.vbank.create_def_with_space(
        4,
        AddressSpace::Register,
        0x20,
        &first.0,
    );
    first.0.write().unwrap().output = Some(first_out.clone());
    let middle_out = fd.vbank.create_def_with_space(
        4,
        AddressSpace::Register,
        0x20,
        &middle.0,
    );
    middle.0.write().unwrap().output = Some(middle_out.clone());

    let first_identity: SeqNum = first.0.read().unwrap().start;
    let middle_identity: SeqNum = middle.0.read().unwrap().start;
    let first_time = first_identity.get_time();
    let middle_time = middle_identity.get_time();
    let same_time_other_address = SeqNum::new(Address::new(0x2000), first_time);
    fd.op_insert_end(&first, &block);
    fd.op_insert_end(&middle, &block);
    fd.op_insert_end(&last, &block);

    for _ in 0..70 {
        let padding = fd.new_op(0, Address::new(0x1000));
        fd.op_set_opcode(&padding, OpCode::CPUI_COPY);
        fd.op_insert_before(&padding, &middle);
    }
    fd.op_uninsert(&first);
    fd.op_insert_end(&first, &block);

    let first_seq = first.0.read().unwrap().start;
    let middle_seq = middle.0.read().unwrap().start;
    let first_lookup = fd.obank.find_op(&first_identity).unwrap();
    let middle_lookup = fd.obank.find_op(&middle_identity).unwrap();
    let first_vn = fd
        .vbank
        .find_vn(4, Address::new(0x20), Address::new(0x1000), first_time)
        .unwrap();
    let middle_vn = fd
        .vbank
        .find_vn(4, Address::new(0x20), Address::new(0x1000), middle_time)
        .unwrap();
    let optree = fd
        .obank
        .optree
        .iter()
        .map(|op| op.0.read().unwrap().get_time().to_string())
        .collect::<Vec<_>>()
        .join(",");

    println!(
        "time_order={}:{},{}:{}",
        first_seq.get_time(),
        first_seq.get_order(),
        middle_seq.get_time(),
        middle_seq.get_order()
    );
    println!(
        "identity={},{}",
        u8::from(first_identity == first_seq),
        u8::from(middle_identity == middle_seq)
    );
    println!(
        "cross_address={},{}",
        u8::from(first_identity.same_identity(&same_time_other_address)),
        u8::from(first_identity < same_time_other_address)
    );
    println!(
        "lookup={},{}",
        u8::from(Arc::ptr_eq(&first_lookup.0, &first.0)),
        u8::from(Arc::ptr_eq(&middle_lookup.0, &middle.0))
    );
    println!(
        "varnode_lookup={},{}",
        u8::from(Arc::ptr_eq(&first_vn, &first_out)),
        u8::from(Arc::ptr_eq(&middle_vn, &middle_out))
    );
    println!("optree_prefix={}", &optree[..5]);
    println!(
        "order_relation={}",
        u8::from(first_seq.get_order() < middle_seq.get_order())
    );

    let graph_dump = dump_dataflow_graph_string(&fd);
    let op_start = graph_dump.find("//START:opnodes").unwrap();
    let op_end = graph_dump[op_start..].find("*END_COLUMNS").unwrap() + op_start;
    let op_section = &graph_dump[op_start..op_end];
    println!(
        "graph_time={},{}",
        u8::from(op_section.contains(&format!("\no{} ", first_time))),
        u8::from(graph_dump.contains(&format!("\no{} v", first_time)))
    );
    let parent_type = Datatype::Base(TypeBase::new(
        "xunknown4".to_string(),
        4,
        TypeMetatype::Unknown,
    ));
    let first_guard = first.0.read().unwrap();
    let middle_guard = middle.0.read().unwrap();
    let first_slot_one = ResolveEdge::new(&parent_type, &first_guard, 1);
    let middle_slot_zero = ResolveEdge::new(&parent_type, &middle_guard, 0);
    let first_slot_zero = ResolveEdge::new(&parent_type, &first_guard, 0);
    println!(
        "resolve_order={},{}",
        u8::from(first_slot_one < middle_slot_zero),
        u8::from(first_slot_zero < middle_slot_zero)
    );
    drop(first_guard);
    drop(middle_guard);
    println!(
        "compare_name={},{}",
        u8::from(HighVariable::compare_name(
            &first_out.read().unwrap(),
            &middle_out.read().unwrap()
        )),
        u8::from(HighVariable::compare_name(
            &middle_out.read().unwrap(),
            &first_out.read().unwrap()
        ))
    );
    fd.op_destroy(&first);
    println!(
        "destroy_identity={},{}",
        u8::from(
            fd.vbank
                .find_vn(4, Address::new(0x20), Address::new(0x1000), first_time)
                .is_none()
        ),
        u8::from(
            fd.vbank
                .find_vn(4, Address::new(0x20), Address::new(0x1000), middle_time)
                .is_some_and(|found| Arc::ptr_eq(&found, &middle_out))
        )
    );
}
